File: apps/ios/LanjingQuiz/Support/CookieCloudCrypto.swift
import CommonCrypto
import CryptoKit
import Foundation

/// Swift port of apps/web/lib/cookiecloud.js: encryption compatible with the
/// official CookieCloud browser extension (github.com/easychen/CookieCloud).
///
///   the_key = MD5(`${uuid}-${password}`).hex.prefix(16)   // 16-char string
///
///   legacy:   OpenSSL EVP_BytesToKey (MD5) -> AES-256-CBC, random 8-byte
///             salt, PKCS7; payload is base64("Salted__" + salt + ciphertext)
///   fixed:    AES-128-CBC, key = UTF-8 bytes of the_key, zero IV, PKCS7;
///             payload is raw base64 ciphertext
///
/// We always push with "aes-128-cbc-fixed" and decrypt either type. Decryption
/// fails closed: a missing "Salted__" header, an unknown crypto type or a bad
/// password (PKCS7 padding error) throws instead of returning garbage.
enum CookieCloudCrypto {
    enum Error: Swift.Error, LocalizedError {
        case unsupportedCryptoType(String)
        case missingSaltedHeader
        case invalidCiphertext
        case invalidPlaintext

        var errorDescription: String? {
            switch self {
            case .unsupportedCryptoType(let type): "cookiecloud: unsupported crypto_type \(type)"
            case .missingSaltedHeader: "cookiecloud: legacy payload is missing the Salted__ header"
            case .invalidCiphertext: "cookiecloud: failed to decrypt (bad password or payload)"
            case .invalidPlaintext: "cookiecloud: decrypted payload is not valid UTF-8"
            }
        }
    }

    private static let saltedPrefix = Data("Salted__".utf8)

    static func deriveKey(uuid: String, password: String) -> String {
        String(Hashing.md5Hex("\(uuid)-\(password)").prefix(16))
    }

    /// OpenSSL EVP_BytesToKey with MD5: D1=MD5(pass+salt), D2=MD5(D1+pass+salt),
    /// D3=MD5(D2+pass+salt); key = D1||D2 (32 B), iv = D3 (16 B).
    private static func evpBytesToKey(passphrase: String, salt: Data) -> (key: Data, iv: Data) {
        let pass = Data(passphrase.utf8)
        var previous = Data()
        var material = Data()
        while material.count < 48 {
            let digest = Insecure.MD5.hash(data: previous + pass + salt)
            let block = Data(digest)
            material.append(block)
            previous = block
        }
        return (key: material.prefix(32), iv: material.subdata(in: 32..<48))
    }

    private static func aesCBC(_ data: Data, key: Data, iv: Data, encrypt: Bool) throws -> Data {
        let operation = CCOperation(encrypt ? kCCEncrypt : kCCDecrypt)
        let capacity = data.count + kCCBlockSizeAES128
        var output = Data(count: capacity)
        var outputLength = 0
        let status = output.withUnsafeMutableBytes { outBytes in
            data.withUnsafeBytes { inBytes in
                key.withUnsafeBytes { keyBytes in
                    iv.withUnsafeBytes { ivBytes in
                        CCCrypt(
                            operation,
                            CCAlgorithm(kCCAlgorithmAES),
                            CCOptions(kCCOptionPKCS7Padding),
                            keyBytes.baseAddress, key.count,
                            ivBytes.baseAddress,
                            inBytes.baseAddress, data.count,
                            outBytes.baseAddress, capacity,
                            &outputLength
                        )
                    }
                }
            }
        }
        guard status == kCCSuccess else { throw Error.invalidCiphertext }
        return output.prefix(outputLength)
    }

    private static func randomSalt() -> Data {
        Data((0..<8).map { _ in UInt8.random(in: .min ... .max) })
    }

    static func legacyEncrypt(_ data: String, key: String, salt: Data = CookieCloudCrypto.randomSalt()) -> String {
        let derived = evpBytesToKey(passphrase: key, salt: salt)
        let ciphertext = try! aesCBC(Data(data.utf8), key: derived.key, iv: derived.iv, encrypt: true)
        return (saltedPrefix + salt + ciphertext).base64EncodedString()
    }

    static func legacyDecrypt(_ encrypted: String, key: String) throws -> String {
        guard let payload = Data(base64Encoded: encrypted),
              payload.count >= saltedPrefix.count + 8,
              payload.prefix(saltedPrefix.count) == saltedPrefix
        else { throw Error.missingSaltedHeader }
        let salt = payload[saltedPrefix.count..<(saltedPrefix.count + 8)]
        let ciphertext = payload[(saltedPrefix.count + 8)...]
        let derived = evpBytesToKey(passphrase: key, salt: Data(salt))
        let plaintext = try aesCBC(Data(ciphertext), key: derived.key, iv: derived.iv, encrypt: false)
        guard let text = String(data: plaintext, encoding: .utf8) else { throw Error.invalidPlaintext }
        return text
    }

    static func fixedEncrypt(_ data: String, key: String) -> String {
        let ciphertext = try! aesCBC(Data(data.utf8), key: Data(key.utf8), iv: Data(count: 16), encrypt: true)
        return ciphertext.base64EncodedString()
    }

    static func fixedDecrypt(_ encrypted: String, key: String) throws -> String {
        guard let ciphertext = Data(base64Encoded: encrypted) else { throw Error.invalidCiphertext }
        let plaintext = try aesCBC(ciphertext, key: Data(key.utf8), iv: Data(count: 16), encrypt: false)
        guard let text = String(data: plaintext, encoding: .utf8) else { throw Error.invalidPlaintext }
        return text
    }

    static func encrypt(_ data: String, uuid: String, password: String, cryptoType: String = "aes-128-cbc-fixed") throws -> String {
        let key = deriveKey(uuid: uuid, password: password)
        switch cryptoType {
        case "aes-128-cbc-fixed": return fixedEncrypt(data, key: key)
        case "legacy": return legacyEncrypt(data, key: key)
        default: throw Error.unsupportedCryptoType(cryptoType)
        }
    }

    /// Missing crypto_type defaults to "legacy" (the extension's default
    /// writer); anything else fails closed.
    static func decrypt(_ encrypted: String, uuid: String, password: String, cryptoType: String = "legacy") throws -> String {
        let key = deriveKey(uuid: uuid, password: password)
        switch cryptoType {
        case "aes-128-cbc-fixed": return try fixedDecrypt(encrypted, key: key)
        case "legacy": return try legacyDecrypt(encrypted, key: key)
        default: throw Error.unsupportedCryptoType(cryptoType)
        }
    }
}

File: apps/android/LanjingQuiz/src/main/java/com/qzh/lanjingquiz/Support/CookieCloudCrypto.kt
package com.qzh.lanjingquiz.Support

import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.spec.IvParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * apps/web/lib/cookiecloud.js 的 Kotlin 移植:与官方 CookieCloud 浏览器扩展加解密兼容。
 *
 *   the_key = MD5(`${uuid}-${password}`).hex.prefix(16)   // 16 字符 key 串
 *
 *   legacy:   OpenSSL EVP_BytesToKey(MD5) -> AES-256-CBC,随机 8 字节盐,PKCS7;
 *             payload 为 base64("Salted__" + salt + ciphertext)
 *   fixed:    AES-128-CBC,key = the_key 的 UTF-8 字节,零 IV,PKCS7;
 *             payload 为裸 base64 密文
 *
 * 本 app 恒推送 "aes-128-cbc-fixed",解密两种都支持。解密 fail-closed:
 * 缺 "Salted__" 头、未知 crypto_type 或错误密码(PKCS7 padding 错误)抛异常而非返回垃圾。
 */
object CookieCloudCrypto {
    // key = UTF-8 字节 of 前16字符 of MD5("{uuid}-{password}") 小写 hex
    fun deriveKey(uuid: String, password: String): ByteArray {
        val hex = Hashers.md5Hex("$uuid-$password")
        return hex.substring(0, 16).toByteArray(Charsets.UTF_8)
    }

    fun fixedEncrypt(plain: String, key: ByteArray): String {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(ByteArray(16)))
        return Base64.getEncoder().encodeToString(cipher.doFinal(plain.toByteArray(Charsets.UTF_8)))
    }

    fun fixedDecrypt(base64Cipher: String, key: ByteArray): String {
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(ByteArray(16)))
        return String(cipher.doFinal(Base64.getDecoder().decode(base64Cipher)), Charsets.UTF_8)
    }

    /**
     * legacy: base64(Salted__ + 8B salt + AES-256-CBC(PKCS7), key=32B, iv=16B via EVP_BytesToKey MD5)。
     * EVP_BytesToKey 的 passphrase 是派生出的 16 字符 key 串(iOS/扩展/web 均如此),
     * 不是原始密码。
     */
    fun legacyDecrypt(base64Cipher: String, uuid: String, password: String): String {
        val full = Base64.getDecoder().decode(base64Cipher)
        if (full.size < 16 || !String(full, 0, 8, Charsets.US_ASCII).equals("Salted__")) {
            throw IllegalArgumentException("missingSaltedHeader")
        }
        val salt = full.copyOfRange(8, 16)
        val (key, iv) = evpBytesToKey(deriveKey(uuid, password), salt, 48)
        val cipher = Cipher.getInstance("AES/CBC/PKCS5Padding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), IvParameterSpec(iv))
        return String(cipher.doFinal(full.copyOfRange(16, full.size)), Charsets.UTF_8)
    }

    /** EVP_BytesToKey MD5:D_i = MD5(D_{i-1} + passphrase + salt);key=前32B, iv=次16B。 */
    private fun evpBytesToKey(passphrase: ByteArray, salt: ByteArray, total: Int): Pair<ByteArray, ByteArray> {
        val md = java.security.MessageDigest.getInstance("MD5")
        var prev = ByteArray(0)
        val out = java.io.ByteArrayOutputStream()
        while (out.size() < total) {
            md.reset()
            md.update(prev)
            md.update(passphrase)
            md.update(salt)
            prev = md.digest()
            out.write(prev)
        }
        val all = out.toByteArray()
        return all.copyOfRange(0, 32) to all.copyOfRange(32, 48)
    }

    fun encryptAny(plain: String, uuid: String, password: String): Pair<String, String> {
        val key = deriveKey(uuid, password)
        return fixedEncrypt(plain, key) to "aes-128-cbc-fixed"
    }

    /** 先按声明类型,失败再试另一算法;双失败抛最后一次错误。 */
    fun decryptAny(encrypted: String, uuid: String, password: String, cryptoType: String?): String {
        val key = deriveKey(uuid, password)
        val errors = ArrayList<Throwable>()
        val attempt: (() -> String) = when (cryptoType ?: "legacy") {
            "aes-128-cbc-fixed" -> ({ fixedDecrypt(encrypted, key) })
            else -> ({ legacyDecrypt(encrypted, uuid, password) })
        }
        try { return attempt() } catch (e: Throwable) { errors.add(e) }
        try {
            return if (cryptoType == "aes-128-cbc-fixed") legacyDecrypt(encrypted, uuid, password)
                   else fixedDecrypt(encrypted, key)
        } catch (e: Throwable) { errors.add(e) }
        throw errors.last()
    }
}
